coord: care about since in determine_frontier

mjibson opened the pull request:
Previously determine_frontier had branching logic that would care about
sinces or uppers based on the presence of an index. I'm not aware of
a reason for this distinction. If there was an index, only the upper
was used in determining a timestamp. We do not guarantee that upper >
since, and so sinks could choose a timestamp < since if the chosen
index hadn't had update_upper run and do its normal "since = upper -
compaction_window" thing. Fix this by consulting sinces and use the
earliest.

Also care about table linearizability while here.

Fixes #6687

benesch commented:
I did some archaeology. As best as I can tell, the intent was that if you tailed a thing that was materialized, you'd only see updates starting from now, whereas if you tailed a thing that was unmaterialized you'd necessarily see the history from the beginning of time (because with RT sources time effectively doesn't start until you choose to materialize the sources). That's not actually what the code did, nor was it the most coherent policy.

I do worry a bit that "TAIL starts from the lagging edge of the compaction window" is also a weird policy call. In a perfect world I feel like

```
INSERT INTO t ...
TAIL t
```

would _not_ show you that insert because it happened before the TAIL started. With this change, IIUC, provided the compaction window is large enough you will see that insert in the TAIL stream in the past. 

Concretely, I'm wondering if `determine_frontier` should use `greatest_open_upper` instead of `least_valid_since`? 

In any case I'm now officially LGTM on removing the branch, provided we sort out the policy question.

mjibson commented:
Yeah, I think it makes since to care about uppers and INSERT linearizability. RIP the 2 line function though :(. Added all the complexity back + more - the branch. PTAL.

benesch commented:
Code LGTM, but I'm nervous that we don't have a test. Can that Go program be whipped up into a Rust test?

mjibson commented:
Tests were failing because we can only use uppers if the indexes are complete, otherwise we'd use max_value, which is also wrong. Added a third branch to do this, which also sheds light on why things existed in the way they did before. At least it's better documented why they are now.

I added a testdrive version of the test. But since it's inherently a race condition bug, the test is (well, was before this fix) flakey. Better than nothing I guess and it will probably start flaking in the future if we break this again.

PTAL again again? Want to make sure this is right.

mjibson commented:
Pushed yet another update that also checks for at least 1 index so we don't use max in the 0 index (static views) case. This seems to be approaching correct.

frankmcsherry reviewed (approved):
So, I have a maybe odd take on this. The new code which advances the proposal by `since` is good, and we want to have it. The rest of the stuff smell like policy, and tbh I don't grok all of it / could imagine it changing. At the same time, I don't really have any strong feelings about it. But I do have strong feelings about the `since` line, which I like.

Is there a near future where we can partition logic into "this part is policy, and 100% up for negotiation" and "this part is mandatory, and not up for negotiation"? It might involve a re-dangling of the coordinator, and I don't want to make work. But at the same time, there are probably different reasons / stakeholders for different parts of this PR.

There's a `println!` still in there, but as long as that gets ripped out I'm good.

frankmcsherry commented:
Regarding linearizability, does some logic somewhere else observe that a table participated in the view, and in the case that `since` advances the query timestamp then bump the coordinator timestamp too? Or do SINK and TAIL not forcibly advance coordinator time (which seems reasonable, as long as they are timestamped in the output).

frankmcsherry commented:
> hopefully he remembers why the branch was added in the first place.

Nope. Though I do believe this code was written in the days before it was acceptable to block on a query and not start streaming results immediately. It was also revised a few times, iirc, which might suggest that we hadn't really nailed it.

I have to admit, I don't grok the current policy either. I think my brain approaches this slightly differently, starting from a "for what time interval can we produce valid data" and then subsetting that down based on policy (e.g. "let's pick a most recent time not in advance of the coordinator time"). I wonder if something like that might result in clearer separation of concerns?

mjibson commented:
Removed the `println`.

I appreciate the feedback to try to separate concerns. I will tinker with a followup PR that attempts to clearly separate the policy from the requirements and see if that improves things.

mjibson commented:
Re linearizability: the special table handling code here is identical to how peeks are handled for tables. If a write happens to any table, then any future peek (and now sink/tail) will use the wall clock time to ensure the write is observed. None of the table time logic cares about `sinces` or `uppers` (although they are still tracked). Tables are super special and have their own timestamp handling using a few extra fields on the coordinator. (Does this answer your question? I couldn't quite parse what it was asking.)

I would love to make tables just another regular source and remove some of the special handling in the coordinator but haven't figured out how yet.

frankmcsherry commented:
> Does this answer your question?

I think so. Or, I think I forgot that when we do a read at a future system time, we don't feel especially bad. Elsewhere, we just make sure the SELECT doesn't come back until that time is ready (I think) and here, .. I guess we produce output data with the right timestamp, so if it starts spilling out (with that timestamp) before the system has reached those timestamps, it isn't the end of the world.

I guess the "bad case" idk about is a user starts a SINK / TAIL and observes a record at a future time, one that the coordinator hasn't reached yet (idk how; constant collection, some coarse ts read granularity), they could plausibly return to MZ and try to SELECT the record but not be allowed to see it yet (if the coordinator's timestamp is not as far along as data that the sink is producing).

mjibson commented:
> they could plausibly return to MZ and try to SELECT the record but not be allowed to see it yet

I can't think of a way this would happen. If a sink/tail produces a record at a ts, then they should be able to SELECT AS OF <ts> and immediately see it (assuming it hasn't been compacted away). There's not a "coordinator timestamp". The coordinator uses wall time only when dealing with tables to ensure that if a table write happens at X all future table reads are >X, but otherwise doesn't care about wall time and has no concept of now or a timestamp.
